File: ui/ui-commons/src/main/java/fr/gouv/vitamui/ui/commons/service/ApplicationService.java
/**
 * Copyright French Prime minister Office/SGMAP/DINSIC/Vitam Program (2019-2020)
 * and the signatories of the "VITAM - Accord du Contributeur" agreement.
 *
 * contact@programmevitam.fr
 *
 * This software is a computer program whose purpose is to implement
 * implement a digital archiving front-office system for the secure and
 * efficient high volumetry VITAM solution.
 *
 * This software is governed by the CeCILL-C license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL-C
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C license and that you accept its terms.
 */
package fr.gouv.vitamui.ui.commons.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import fr.gouv.vitamui.commons.api.CommonConstants;
import fr.gouv.vitamui.commons.api.domain.*;
import fr.gouv.vitamui.commons.api.enums.AttachmentType;
import fr.gouv.vitamui.commons.api.logger.VitamUILogger;
import fr.gouv.vitamui.commons.api.logger.VitamUILoggerFactory;
import fr.gouv.vitamui.commons.rest.client.ExternalHttpContext;
import fr.gouv.vitamui.commons.security.client.logout.CasLogoutUrl;
import fr.gouv.vitamui.iam.external.client.ApplicationExternalRestClient;
import fr.gouv.vitamui.iam.external.client.IamExternalRestClientFactory;
import fr.gouv.vitamui.ui.commons.config.AutoConfigurationVitam;
import fr.gouv.vitamui.ui.commons.property.PortalCategoryConfig;
import fr.gouv.vitamui.ui.commons.property.UIProperties;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.boot.info.BuildProperties;

import javax.validation.constraints.NotNull;
import javax.xml.bind.DatatypeConverter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 *
 */
@EnableConfigurationProperties
@ConfigurationProperties
public class ApplicationService extends AbstractCrudService<ApplicationDto> {

    private static final VitamUILogger LOGGER = VitamUILoggerFactory.getInstance(ApplicationService.class);

    private static final String DELIMITER = ".";

    private static final String VERSION_RELEASE_KEY = "version.release";

    private static String PLATFORM_NAME = "PLATFORM_NAME";

    private final UIProperties properties;

    private final ApplicationExternalRestClient client;

    private final CasLogoutUrl casLogoutUrl;

    private final BuildProperties buildProperties;

    @Autowired
    private AutoConfigurationVitam autoConfigurationVitam;

    @Value("${cas.external-url}")
    @NotNull
    private String casExternalUrl;

    @Value("${cas.callback-url}")
    @NotNull
    private String casCallbackUrl;

    @Value("${ui.url}")
    @NotNull
    private String uiUrl;

    @Value("${ui.redirect-url}")
    @NotNull
    private String uiRedirectUrl;

    private Map<String, List<String>> listEnableExternalIdentifiers;

    public ApplicationService(final UIProperties properties, final CasLogoutUrl casLogoutUrl, final IamExternalRestClientFactory factory,
    final BuildProperties buildProperties) {
        this.properties = properties;
        this.casLogoutUrl = casLogoutUrl;
        this.buildProperties = buildProperties;
        if (this.properties == null) {
            LOGGER.warn("Properties not provided");
        } else if (this.properties.getBaseUrl() == null) {
            LOGGER.warn("base-url properties not provided");
        }
        client = factory.getApplicationExternalRestClient();
    }

    public Map<String, Object> getApplications(final ExternalHttpContext context, final boolean filterApp) {
        final QueryDto query = new QueryDto(QueryOperator.AND);
        query.addCriterion(new Criterion("filterApp", filterApp, CriterionOperator.EQUALS));

        Collection<ApplicationDto> applications = client.getAll(context, Optional.of(query.toJson()));
        Map<String, PortalCategoryConfig> categories = properties.getPortalCategories();
        categories.keySet().forEach(category -> categories.get(category).setIdentifier(category));

        Map<String, Object> portalConfig = new HashMap<>();
        portalConfig.put(CommonConstants.APPLICATION_CONFIGURATION, applications);
        portalConfig.put(CommonConstants.CATEGORY_CONFIGURATION, categories);

        return portalConfig;
    }

    private Map<String, List<String>> getListEnableExternalIdentifiers() {
        if(listEnableExternalIdentifiers == null) {
            listEnableExternalIdentifiers = autoConfigurationVitam.getTenants();
        }
        return listEnableExternalIdentifiers;
    }

    public Boolean isApplicationExternalIdentifierEnabled(final ExternalHttpContext context, final String identifier) {
        final String tenantId = context.getTenantIdentifier().toString();
        final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());

        final Map<String, List<String>> externalIdentifiers = getListEnableExternalIdentifiers();
        LOGGER.info("Reading list of external identifiers {}", externalIdentifiers == null ? "null" : externalIdentifiers.toString());

        if(externalIdentifiers != null) {
            if(listEnableExternalIdentifiers.containsKey(tenantId)) {
                final List<String> enabledApplications = listEnableExternalIdentifiers.get(tenantId);
                return(enabledApplications.contains(identifier));
            }
        }

        return false;
    }

    public String getBaseUrlPortal() {
        return properties.getBaseUrl().getPortal();
    }

    public Map<String, Object> getConf() {
        final Map<String, Object> configurationData = new HashMap<>();
        configurationData.put(CommonConstants.PORTAL_URL, properties.getBaseUrl().getPortal());
        configurationData.put(CommonConstants.CAS_LOGIN_URL, getCasLoginUrl());
        configurationData.put(CommonConstants.CAS_LOGOUT_URL, casLogoutUrl.getValue());
        configurationData.put(CommonConstants.UI_URL, uiUrl);
        configurationData.put(CommonConstants.LOGOUT_REDIRECT_UI_URL, casLogoutUrl.getValueWithRedirection(uiRedirectUrl));
        configurationData.put(CommonConstants.THEME_COLORS, properties.getThemeColors());
        configurationData.put(CommonConstants.PORTAL_TITLE, properties.getPortalTitle());
        configurationData.put(CommonConstants.PORTAL_MESSAGE, properties.getPortalMessage());
        configurationData.put(CommonConstants.CUSTOMER, properties.getCustomer());
        String versionRelease = properties.getVersionRelease();
        if (StringUtils.isEmpty(versionRelease)) {
            versionRelease = Stream.of(buildProperties.get(VERSION_RELEASE_KEY).split("\\" + DELIMITER)).limit(2).map(Object::toString)
                    .collect(Collectors.joining(DELIMITER));

        }
        if (StringUtils.isNotEmpty(versionRelease)) {
            configurationData.put(CommonConstants.VERSION_RELEASE, versionRelease);
        }
        if (properties.getPlatformName() != null) {
            configurationData.put(PLATFORM_NAME, properties.getPlatformName());
        } else {
            configurationData.put(PLATFORM_NAME, "VITAM-UI");
        }
        return configurationData;
    }

    private String getCasLoginUrl() {
        return casExternalUrl + "/login?service=" + casCallbackUrl;
    }


    @Override
    public ApplicationExternalRestClient getClient() {
        return client;
    }

    public String getBase64File(final String fileName, final String basePath) {
        if (StringUtils.isBlank(fileName) || StringUtils.isBlank(basePath)) {
            LOGGER.warn(String.format("Logo information missing : cannot load logo with name \"%s\" in path \"%s\"", fileName, basePath));
            return null;
        }

        final Path assetFile = Paths.get(basePath, Paths.get(fileName).getFileName().toString());
        String base64Asset = null;
        try {
            base64Asset = DatatypeConverter.printBase64Binary(Files.readAllBytes(assetFile));
        } catch (final IOException e) {
            LOGGER.error("Error while resolving logo path", e);
            return null;
        }
        return base64Asset;
    }

    public Map<String, Object> getBase64Assets(final List<AttachmentType> assets) {
        final Map<String, Object> files = new HashMap<>();
        assets.forEach(asset -> {
            String file = null;
            switch (asset) {
                case HEADER:
                    file = getBase64File(properties.getHeaderLogo(), properties.getAssets());
                    break;
                case FOOTER:
                    file = getBase64File(properties.getFooterLogo(), properties.getAssets());
                    break;
                case PORTAL:
                    file = getBase64File(properties.getPortalLogo(), properties.getAssets());
                    break;
            }
            files.put(asset.value(), file);
        });
        return files;
    }

}

File: ui/ui-frontend/projects/vitamui-library/src/app/modules/application.service.ts
/*
 * Copyright French Prime minister Office/SGMAP/DINSIC/Vitam Program (2019-2022)
 * and the signatories of the "VITAM - Accord du Contributeur" agreement.
 *
 * contact@programmevitam.fr
 *
 * This software is a computer program whose purpose is to implement
 * implement a digital archiving front-office system for the secure and
 * efficient high volumetry VITAM solution.
 *
 * This software is governed by the CeCILL-C license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL-C
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C license and that you accept its terms.
 */
import { Injectable, inject } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject, Observable, of } from 'rxjs';
import { catchError, filter, map, mergeMap, take, tap } from 'rxjs/operators';
import { ApplicationApiService } from './api/application-api.service';
import { ApplicationId } from './application-id.enum';
import { AuthService } from './auth.service';
import { ConfigService } from './config.service';
import { GlobalEventService } from './global-event.service';
import { Application } from './models/application/application.interface';
import { Category } from './models/application/category.interface';
import { Tenant } from './models/customer/tenant.interface';
import { ApplicationAnalytics } from './models/user/application-analytics.interface';
import { TenantSelectionService } from './tenant-selection.service';

@Injectable({
  providedIn: 'root',
})
export class ApplicationService {
  private applicationApi = inject(ApplicationApiService);
  private authService = inject(AuthService);
  private tenantService = inject(TenantSelectionService);
  private globalEventService = inject(GlobalEventService);
  private configService = inject(ConfigService);

  set applications(apps: Application[]) {
    this._applications = apps;
    this._applications$.next(this._applications);
  }

  get applicationsAnalytics(): ApplicationAnalytics[] {
    return this._applicationsAnalytics;
  }

  set applicationsAnalytics(apps: ApplicationAnalytics[]) {
    this._applicationsAnalytics = apps;
  }

  get categories(): Category[] {
    return this._categories;
  }

  set categories(categories: Category[]) {
    this._categories = categories;
  }

  private _categories: Category[];
  private _applications: Application[];
  private _applications$ = new BehaviorSubject<Application[]>(null);
  private _applicationsAnalytics: ApplicationAnalytics[];
  private appMap$ = new BehaviorSubject(null);

  /**
   * Get and init applications list for the current auth user.
   */
  public list(): Observable<Application[]> {
    return this.applicationApi.getAll().pipe(
      catchError(() => of([])),
      map((applications: Application[]) => {
        this._applications = applications;
        this._categories = this.sortCategories(this.configService.config.CATEGORY_CONFIGURATION);
        this._applications$.next(this._applications);
        return applications;
      }),
    );
  }

  /**
   * Get Applications list grouped by categories in a hashMap of the active tenant.
   */
  public getActiveTenantAppsMap(): Observable<Map<Category, Application[]>> {
    return this.tenantService.getSelectedTenant$().pipe(
      mergeMap((tenant: Tenant) => this.getTenantAppMap(tenant)),
      tap((appMap: Map<Category, Application[]>) => this.appMap$.next(appMap)),
    );
  }

  /**
   * Returns the provided tenant application map as Map<Category, Application[]>
   * @param tenant - tenant whitch we want applications
   */
  public getTenantAppMap(tenant: Tenant): Observable<Map<Category, Application[]>> {
    return this.getApplications$().pipe(
      map((applications: Application[]) => {
        const apps: Application[] = [];
        const tenantsByApp = this.authService.user.tenantsByApp;
        if (tenantsByApp && tenant) {
          tenantsByApp.forEach((tenantByAppItem: { name: string; tenants: Tenant[] }) => {
            const appTenant = tenantByAppItem.tenants.find((value) => value.identifier === tenant.identifier);
            const app = applications.find((value) => value.identifier === tenantByAppItem.name);
            if (app && (appTenant || !app.hasTenantList)) {
              apps.push(app);
            }
          });

          const resultMap = this.fillCategoriesWithApps(this.categories, apps);
          const lastUsedApps = this.getLastUsedApps(this.categories, apps);

          if (lastUsedApps.category) {
            resultMap.set(lastUsedApps.category.identifier, lastUsedApps.apps);
          }

          const convertedMap = this.convertToCategoryMap(resultMap);
          return this.sortMapByCategory(convertedMap);
        }

        return new Map<Category, Application[]>();
      }),
    );
  }

  public openApplication(app: Application, router: Router, uiUrl: string, tenantId?: number): void {
    this.tenantService.saveTenantIdentifier(tenantId).subscribe(async (savedTenantId: number) => {
      const isSameService = app.serviceId.includes(uiUrl);
      const isHomePage = router.url === '/';
      const hasTenant = app.hasTenantList;
      const baseUrl = app.url.replace(uiUrl, '');

      if (isSameService && !isHomePage) {
        const path = hasTenant ? [baseUrl, 'tenant', savedTenantId] : [baseUrl];
        await router.navigate(path);
      } else {
        window.location.href = hasTenant ? `${app.url}/tenant/${savedTenantId}` : app.url;
      }
    });
  }

  public getUrl$({ appId, tenantIdentifier }: { appId: ApplicationId; tenantIdentifier?: number }): Observable<string> {
    return this.getAppById(appId).pipe(map((app) => this.getApplicationUrl(app, tenantIdentifier)));
  }

  public getApplicationUrl(app: Application, tenantIdentifier?: number): string {
    if (!tenantIdentifier) {
      tenantIdentifier = this.tenantService.getSelectedTenant().identifier;
    }

    if (app.hasTenantList) {
      return app.url + '/tenant/' + tenantIdentifier;
    } else {
      return app.url;
    }
  }

  public getApplicationTenants(appId: string): Tenant[] {
    if (this.authService.user) {
      const appTenantsInfo = this.authService.user.tenantsByApp.find((appTenantInfo) => appTenantInfo.name === appId);
      const appTenants = appTenantsInfo ? appTenantsInfo.tenants : [];
      appTenants.sort((t1, t2) => t1.name.localeCompare(t2.name));
      return appTenants;
    }

    return [];
  }

  public getAppById(identifier: string): Observable<Application> {
    return this.getApplications$().pipe(map((apps: Application[]) => apps.find((value) => value.identifier === identifier)));
  }

  /**
   * Return an observable that notify if the current application has a tenant list or not.
   */
  public hasTenantList(): Observable<boolean> {
    return this.globalEventService.pageEvent.pipe(
      mergeMap((appId: string) => {
        return this.getAppById(appId).pipe(
          map((app: Application) => {
            if (appId === ApplicationId.PORTAL_APP) {
              return this.configService.config.UI?.hasTenantList ?? true;
            }
            return app ? app.hasTenantList : false;
          }),
        );
      }),
    );
  }

  public isApplicationExternalIdentifierEnabled(id: string): Observable<boolean> {
    return this.applicationApi.isApplicationExternalIdentifierEnabled(id).pipe(
      catchError(() => of([])),
      map((result: boolean) => {
        return result;
      }),
    );
  }

  public getApplications$(): Observable<Application[]> {
    return this._applications$.pipe(
      filter((apps: Application[]) => !!apps),
      take(1),
    );
  }

  /**
   * Convert a map using a string category identifier as key to a map using a Category object instead
   * @param stringMap - the map to convert as Map<string, Application[]>
   */
  private convertToCategoryMap(stringMap: Map<string, Application[]>): Map<Category, Application[]> {
    const categMap = new Map<Category, Application[]>();
    stringMap.forEach((val, key) => {
      const categ = this.categories.find((value) => value.identifier === key);
      categMap.set(categ, val);
    });
    return categMap;
  }

  private sortMapByCategory(appMap: Map<Category, Application[]>): Map<Category, Application[]> {
    return new Map([...appMap.entries()].sort((a, b) => (a[0].order < b[0].order ? -1 : 1)));
  }

  private fillCategoriesWithApps(categories: Category[], applications: Application[]): Map<string, Application[]> {
    const resultMap = new Map<string, Application[]>();
    categories.forEach((category: Category) => {
      const sortedAppsOfCategory = this.getSortedAppsOfCategory(category, applications);
      if (sortedAppsOfCategory && sortedAppsOfCategory.length > 0) {
        resultMap.set(category.identifier, sortedAppsOfCategory);
      }
    });
    return resultMap;
  }

  private getLastUsedApps(
    categories: Category[],
    applications: Application[],
    max = 8,
  ): {
    category: Category;
    apps: Application[];
  } {
    let dataSource: ApplicationAnalytics[];
    if (this.applicationsAnalytics) {
      dataSource = this.applicationsAnalytics;
    } else if (this.authService.user.analytics && this.authService.user.analytics.applications) {
      dataSource = this.authService.user.analytics.applications;
    }

    if (dataSource) {
      const lastUsedAppsCateg = {
        order: 0,
        identifier: 'lastusedapps',
        title: 'Dernières utilisées',
        displayTitle: true,
      };
      // Define & set last used apps array
      let lastUsedApps = applications.filter((application: Application) => {
        return dataSource.findIndex((app: ApplicationAnalytics) => app.applicationId === application.identifier) !== -1;
      });

      if (lastUsedApps.length !== 0) {
        // Check if category already exists
        const categoryIndex = categories.findIndex((category: Category) => category.identifier === lastUsedAppsCateg.identifier);
        if (categoryIndex === -1) {
          this.categories.push(lastUsedAppsCateg);
        }

        // Sort last used apps by date
        lastUsedApps.sort((a: Application, b: Application) => {
          const c = dataSource.find((app: ApplicationAnalytics) => app.applicationId === a.identifier);
          const d = dataSource.find((app: ApplicationAnalytics) => app.applicationId === b.identifier);
          return c && d && new Date(c.lastAccess).getTime() > new Date(d.lastAccess).getTime() ? -1 : 1;
        });

        // Get 8 last used apps if there is more than 8
        if (lastUsedApps.length > max) {
          lastUsedApps = lastUsedApps.slice(0, 8);
        }

        return { category: lastUsedAppsCateg, apps: lastUsedApps };
      }
    }

    return { category: undefined, apps: [] };
  }

  private getSortedAppsOfCategory(category: Category, applications: Application[]): Application[] {
    if (applications) {
      const apps = applications.filter((application: Application) => application.category === category.identifier) as Application[];
      return this.sortApplications(apps);
    }

    return [];
  }

  private sortApplications(applications: Application[]): Application[] {
    // Sort apps inside categories
    return applications.sort((a: Application, b: Application) => {
      return a.position < b.position ? -1 : 1;
    });
  }

  private sortCategories(categories: Category[]): Category[] {
    // Sort apps inside categories
    return categories.sort((a, b) => {
      return a.order > b.order ? 1 : -1;
    });
  }
}
